freshmaker/handlers/botas/botas_shipped_advisory.py

```diff
@@ -96,63 +135,144 @@ def handle(self, event):
+        # auto_rebuild_tags of repository and bundle's tags
+        for digest, bundle_data in bundles_by_digest.items():
+            bundle_nvr = bundle_data['nvr']
+
+            # Images are for different arches, just check against the first image
+            image = bundle_data['images'][0]
+            if self.image_has_auto_rebuild_tag(image):
+                bundle_data['auto_rebuild'] = True
+
+            # Fetch buildinfo
+            buildinfo = koji_api.get_build(bundle_nvr)
+            related_images = (
+                buildinfo.get("extra", {})
+                .get("image", {})
+                .get("operator_manifests", {})
+                .get("related_images", {})
+            )
+            bundle_data['osbs_pinning'] = related_images.get('created_by_osbs', False)
+            # Save the original pullspecs
+            bundle_data['pullspecs'] = related_images.get('pullspecs', [])
+
+        # Digests of bundles to be rebuilt
+        to_rebuild_digests = set()
+
+        # Now for each bundle, replace the original digest with rebuilt
+        # digest (override pullspecs)
+        for digest, bundle_data in bundles_by_digest.items():
+            # Override pullspecs only when auto_rebuild is enabled and OSBS-pinning
+            # mechanism is used.
+            if not (bundle_data['auto_rebuild'] and bundle_data['osbs_pinning']):
                 continue
-            bundle_digests.add(bundle['bundle_path_digest'])
-        bundle_images = self._pyxis.get_images_by_digests(bundle_digests)
 
-        # Filter image nvrs that don't have or never had auto_rebuild tag
-        # in repos, where image is published
-        auto_rebuild_nvrs = self._pyxis.get_auto_rebuild_tagged_images(bundle_images)
+            for pullspec in bundle_data['pullspecs']:
+                # A pullspec item example:
+                # {
+                #   'new': 'registry.exampe.io/repo/example-operator@sha256:<sha256-value>'
+                #   'original': 'registry.example.io/repo/example-operator:v2.2.0',
+                #   'pinned': True
+                # }
 
-        # get NVRs only of those bundles, which have OSBS pinning
-        bundles_nvrs = self._filter_bundles_by_pinned_related_images(
-            auto_rebuild_nvrs)
+                # If related image is not pinned by OSBS, skip
+                if not pullspec.get('pinned', False):
+                    continue
+
+                # A pullspec path is in format of "registry/repository@digest"
+                old_pullspec = pullspec.get('new')
+                # Replace 'original' with the old 'new' pullspec
+                pullspec['original'] = old_pullspec
+
+                pullspec_elems = old_pullspec.split('@')
+                old_digest = pullspec_elems[1]
+
+                if old_digest not in original_nvrs_by_digest:
+                    # This related image is not one of the original images
+                    continue
```

Comment:
This check should happen before `pullspec['original'] = old_pullspec` is called since we want the original pullspec replacement that OSBS computed to remain for related images which weren't rebuilt.